### PR TITLE
restic forget --keep-* options will interpret -1 as "forever"

### LINKED ISSUE #2565: restic forget --keep-* options must interpret -1 as "forever".
Output of `restic version`
--------------------------
`restic 0.9.6 compiled with go1.13.4 on linux/amd64`

How did you run restic exactly?
-------------------------------
`restic -r swift:backup:/aeus/restic forget --keep-weekly -1 --keep-within 2m`

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
`openstack swift`

Expected behavior
-----------------
Keep all snapshots within 2 months, keep once per week for earlier period.

Actual behavior
---------------
Keep all snapshots within 2 months. Delete all from the earlier period.

Do you have an idea how to solve the issue?
-------------------------------------------
I'm coming from borg, and found that most options match (which is nice). The `forget` command clearly match `borg prune` one.
But when it comes to say `keep one snapshot per hour/day/week/month`, borg support the value `-1` which is far better than hardcoding a "max" like  --keep-weekly 9999.
I'm quite sad that, while I relied  on the same behavior, `restic` not only ran the cleanup but dropped all my previous backups by misinterpreting `-1`.
At the **very** least, `restic` must bail-out and warn about unknown value.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
It forgot about all my backups older than 2 months because I (apparently mistakenly) typed `--keep-weekly -1 --keep-within 2m`

### PR DESCRIPTION
**Pull request by thndrbrrr**
What does this PR change? What problem does it solve?
-----------------------------------------------------

`restic forget` will interpret `-1` as "forever" when used with `--keep-*` options. Using a value less than `-1` will result in an error.

~~**Note:** This PR does _not_ change the existing behaviour of `--keep-within*` options until it's decided how durations containing negative values should be handled (see my comment in #2565).~~

`--keep-within*` options containing negative values are considered invalid.

Documentation has not been updated (waiting on confirmation for this PR to go ahead).

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

Closes #2565 ~~(if `--keep-within-*` options could be considered a different issue)~~

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

### CODE REVIEW

File: internal/restic/snapshot_policy.go [hunk 1]
```diff
@@ -260,13 +260,16 @@ func ApplyPolicy(list Snapshots, p ExpirePolicy) (keep, remove Snapshots, reason
 
 		// Now update the other buckets and see if they have some counts left.
 		for i, b := range buckets {
-			if b.Count > 0 {
+			if b.Count <= -1 || b.Count > 0 {
```
Comment (MichaelEischer) [inline internal/restic/snapshot_policy.go:263]: `b.Count == -1` should also be enough? (The current condition is just a complicated way of saying `b.Count != 0`.) And please add a comment, that -1 means infinite.

File: internal/restic/snapshot_policy_test.go [hunk 2]
```diff
@@ -251,9 +255,11 @@ func TestApplyPolicy(t *testing.T) {
 					len(keep)+len(remove), len(testExpireSnapshots))
 			}
 
-			if p.Sum() > 0 && len(keep) > p.Sum() {
-				t.Errorf("not enough snapshots removed: policy allows %v snapshots to remain, but ended up with %v",
-					p.Sum(), len(keep))
+			if p.Last >= 0 && p.Hourly >= 0 && p.Daily >= 0 && p.Weekly >= 0 && p.Monthly >= 0 && p.Yearly >= 0 {
+				if p.Sum() > 0 && len(keep) > p.Sum() {
```
Comment (MichaelEischer) [inline internal/restic/snapshot_policy_test.go:259]: `ExpirePolicy.Sum()` is only used by the tests. Thus, please convert that method into a function and move it into the tests. Afterwards, change `Sum()` to return 0 if any of the counts is `-1`.

File: cmd/restic/cmd_forget_test.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+)
+
+func TestPreventNegativeForgetOptionValues(t *testing.T) {
+	invalidForgetOpts := []ForgetOptions{
+		{Last: -2},
+		{Hourly: -2},
+		{Daily: -2},
+		{Weekly: -2},
+		{Monthly: -2},
+		{Yearly: -2},
```
Comment (MichaelEischer) [inline cmd/restic/cmd_forget_test.go:17]: Please also add a basic test to check that -1 and positive counts are accepted.

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
Thanks for working on this issue! The code looks solid except for the comments below. Regarding `--keep-within-*`, I think restic should just return an error if any of them contains a negative duration. I don't see a use case in which that would make sense.

**Comment by thndrbrrr** [comment]
Thanks for the feedback, @MichaelEischer! I'll update the PR with the recommended changes. 👍

File: cmd/restic/cmd_forget.go [hunk 4]
```diff
@@ -99,8 +99,28 @@ func init() {
 	addPruneOptions(cmdForget)
 }
 
+func verifyForgetOptions(opts *ForgetOptions) error {
+	var negValFound = false
+
+	if opts.Last < -1 || opts.Hourly < -1 || opts.Daily < -1 || opts.Weekly < -1 ||
+		opts.Monthly < -1 || opts.Yearly < -1 {
+		negValFound = true
```
[the remark this answers is not in the document]
Comment (thndrbrrr) [reply cmd/restic/cmd_forget.go:107]: Removed `negValFound` and added checks for `--keep-within*` flags

File: cmd/restic/cmd_forget_test.go (hunk 3 above)
Comment (thndrbrrr) [reply cmd/restic/cmd_forget_test.go:17]: Added tests for -1, 0, and positive values, as well as tests for valid and invalid `Within*` options.

**Comment by thndrbrrr** [comment]
@MichaelEischer Alrighty, made the changes. This resulted in minor refactoring: had to move (and slightly change) a helper function from `cmd_snapshots_test.go` into `testing.go`.

**Comment by thndrbrrr** [comment]
I'll update this PR further with changes to `func (e ExpirePolicy) String() (s string)` to make command line output clearer. Right  now an option like `--keep-daily -1` prints `Applying Policy: keep` ... which misses an important point ... 😉

**Comment by thndrbrrr** [comment]
Ok, updated PR. Options such as `--keep-daily 7 --keep-yearly -1 --keep-monthly -1 --keep-hourly 24` for example will now print:
```
Applying Policy: keep 24 hourly, 7 daily, all monthly, all yearly snapshots
```

**Comment by thndrbrrr** [comment]
@MichaelEischer Any feedback on the updated PR? Happy to update the docs if that's all that's misisng at this point.

File: internal/restic/testing.go [hunk 5]
```diff
@@ -212,3 +212,14 @@ func TestParseHandle(s string, t BlobType) BlobHandle {
 func TestSetSnapshotID(t testing.TB, sn *Snapshot, id ID) {
 	sn.id = &id
 }
+
+// Convenience function that parses a duration from a string or panics if string is invalid.
```
[the remark this answers is not in the document]
Comment (thndrbrrr) [reply internal/restic/testing.go]: Done.

**Comment by thndrbrrr** [comment]
@MichaelEischer Docs and changelog updated.

I updated the usage text as well, let me know if you're not happy with the wording:

```
$ ./restic forget -h
[...]
Usage:
  restic forget [flags] [snapshot ID] [...]

Flags:
  -l, --keep-last n                    keep the last n snapshots (use "-1" for n to keep all snapshots)
  -H, --keep-hourly n                  keep the last n hourly snapshots (use "-1" for n to keep all hourly snapshots)
  -d, --keep-daily n                   keep the last n daily snapshots (use "-1" for n to keep all daily snapshots)
  -w, --keep-weekly n                  keep the last n weekly snapshots (use "-1" for n to keep all weekly snapshots)
  -m, --keep-monthly n                 keep the last n monthly snapshots (use "-1" for n to keep all monthly snapshots)
  -y, --keep-yearly n                  keep the last n yearly snapshots (use "-1" for n to keep all yearly snapshots)
```

**Comment by thndrbrrr** [comment]
> @thndrbrrr Sorry for the late reply, I've been far too busy during the last few weeks.

No worries, I know how that feels ... 😆

File: cmd/restic/cmd_forget.go [hunk 6]
```diff
@@ -67,12 +67,12 @@ func init() {
 	cmdRoot.AddCommand(cmdForget)
 
 	f := cmdForget.Flags()
-	f.IntVarP(&forgetOptions.Last, "keep-last", "l", 0, "keep the last `n` snapshots")
-	f.IntVarP(&forgetOptions.Hourly, "keep-hourly", "H", 0, "keep the last `n` hourly snapshots")
-	f.IntVarP(&forgetOptions.Daily, "keep-daily", "d", 0, "keep the last `n` daily snapshots")
-	f.IntVarP(&forgetOptions.Weekly, "keep-weekly", "w", 0, "keep the last `n` weekly snapshots")
-	f.IntVarP(&forgetOptions.Monthly, "keep-monthly", "m", 0, "keep the last `n` monthly snapshots")
-	f.IntVarP(&forgetOptions.Yearly, "keep-yearly", "y", 0, "keep the last `n` yearly snapshots")
+	f.IntVarP(&forgetOptions.Last, "keep-last", "l", 0, "keep the last `n` snapshots (use \"-1\" for n to keep all snapshots)")
```
Comment (MichaelEischer) [inline cmd/restic/cmd_forget.go:70]: My feeling is that `(use \"-1\" to keep all snapshots)` might be easier to understand. After all, there's no other place in which one could add the `-1`.

**Review by MichaelEischer — APPROVED** [review]
I only have one small remaining nitpick. Besides that LGTM.

File: cmd/restic/cmd_forget.go (hunk 6 above)
Comment (thndrbrrr) [reply cmd/restic/cmd_forget.go]: 👍

**Pull request merged**